managedservices/managedservices.go

```diff
@@ -0,0 +1,230 @@
+		}
+
+		resource, err = frameworkresource.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	ms := &ManagedServices{
+		apprClient:    config.ApprClient,
+		helmClient:    config.HelmClient,
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+		resource:      resource,
+	}
+
+	return ms, nil
+}
+
+func (ms *ManagedServices) Test(ctx context.Context, chartConfig ChartConfig, chartResources ChartResources) error {
+	var err error
+
+	err = validateChartConfig(chartConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	err = validateChartResources(chartResources)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("installing chart %#q", chartConfig.ChartName))
+
+		err = ms.resource.InstallResource(chartConfig.ChartName, chartConfig.ChartValues, chartConfig.ChannelName)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("installed chart %#q", chartConfig.ChartName))
+	}
+
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "waiting for deployed status")
+
+		err = ms.resource.WaitForStatus(chartConfig.ChartName, "DEPLOYED")
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "chart is deployed")
+	}
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "checking resources")
+
+		for _, ds := range chartResources.DaemonSets {
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("checking daemonset %#q", ds.Name))
+
+			err = ms.checkDaemonSet(ds)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("daemonset %#q is correct", ds.Name))
+		}
+
+		for _, d := range chartResources.Deployments {
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("checking deployment %#q", d.Name))
+
+			err = ms.checkDeployment(d)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment %#q is correct", d.Name))
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "resources are correct")
+	}
+
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "running release tests")
+
+		err = ms.helmClient.RunReleaseTest(chartConfig.ChartName)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "release tests passed")
+	}
+
+	return nil
+}
+
+// checkDaemonSet ensures that key properties of the daemonset are correct.
+func (ms *ManagedServices) checkDaemonSet(expectedDaemonSet DaemonSet) error {
+	ds, err := ms.hostFramework.K8sClient().Apps().DaemonSets(expectedDaemonSet.Namespace).Get(expectedDaemonSet.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return microerror.Maskf(notFoundError, "daemonset: %#q", expectedDaemonSet.Name, err)
```

Comment:
The format looks wrong. There is only one format placeholder but two arguments.

Reply:
Thanks good catch. I'll fix.